Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -899,28 +895,39 @@ Axis *MatrixWorkspace::getAxis(const std::size_t &axisIndex) const {
         "Argument to getAxis is invalid for this workspace");
   }
 
-  return m_axes[axisIndex];
+  return m_axes[axisIndex].get();
 }
 
 /** Replaces one of the workspace's axes with the new one provided.
  *  @param axisIndex :: The index of the axis to replace
- *  @param newAxis :: A pointer to the new axis. The class will take ownership.
+ *  @param newAxis :: A Unique_ptr to the new axis. The class will take
+ * ownership.
  *  @throw IndexError If the axisIndex given is outside the range of axes held
  * by this workspace
  *  @throw std::runtime_error If the new axis is not of the correct length
  * (within one of the old one)
  */
 void MatrixWorkspace::replaceAxis(const std::size_t &axisIndex,
-                                  Axis *const newAxis) {
+                                  std::unique_ptr<Axis> newAxis) {
   // First check that axisIndex is in range
   if (axisIndex >= m_axes.size()) {
     throw Kernel::Exception::IndexError(
         axisIndex, m_axes.size(),
         "Value of axisIndex is invalid for this workspace");
   }
   // If we're OK, then delete the old axis and set the pointer to the new one
-  delete m_axes[axisIndex];
-  m_axes[axisIndex] = newAxis;
+  m_axes[axisIndex] = std::move(newAxis);
+}
+
+/**
+ * Raw Pointer version of replaceAxis to allow it to work with python
+ *
+ * @param axisIndex :: The index of the axis to replace
+ * @param newAxis :: A pointer to the new axis. The class will take ownership.
+ */
+void MatrixWorkspace::pythonReplaceAxis(const std::size_t &axisIndex,
```

Comment:
This is exactly the correct thing to do but not quite in the right place. The [Python layer](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/api/src/Exports/MatrixWorkspace.cpp) is able to contain these so called "proxy" functions as intermediaries without them needing to clutter up the C++ API.